### PR TITLE
adding the social grade and the personal students preferred students.

### PR DESCRIPTION
**Pull request by noy248**
Student page with full parameters view and editing.
adding the social grading and personal preffered of the students

### CODE REVIEW

File: client/app/students/new-student/new-student.html [hunk 1]
```diff
@@ -5,6 +5,7 @@
         </div>
     </md-toolbar>
     <form name="addStudentForm">
+       <div>
```
Comment (edenbs) [inline client/app/students/new-student/new-student.html:8]: Why did you add this empty div? it doesn't seem to do anything
Delete it if it's unnecessary

File: client/app/students/new-student/new-student.html [hunk 2]
```diff
@@ -33,9 +34,36 @@
...
+                </md-input-container>
+               <div ng-messages="addStudentForm.FirstPrefer.$error" md-auto-hide="true"
+                    ng-if="(addStudentForm.FirstPrefer.$invalid && addStudentForm.SecondPrefer.$invalid )&&
+                     (addStudentForm.FirstPrefer.$touched && addStudentForm.SecondPrefer.$touched)">
+                   <p ng-message="pattern">Personal preferations must be at most two students id </p>
+               </div>
+           </div>
         </md-dialog-contnet>
+       </div>
+        <div>
```
Comment (edenbs) [inline client/app/students/new-student/new-student.html]: Same as above

File: client/app/students/new-student/new-student.html [hunk 3]
```diff
@@ -33,9 +34,36 @@
                     <p ng-message="pattern">Class must be formatted כיתה-מספר</p>
                 </div>
             </md-input-container>
+            <md-input-container>
+                <label>Social Grade</label>
+                <input name="social" ng-model="student.social"  ng-pattern='/^([1-4])$/' required/>
```
Comment (edenbs) [inline client/app/students/new-student/new-student.html]: Maybe this should be `type="number"`? it will automatically convert it to a number

File: server/api/student/student.model.js [hunk 4]
```diff
@@ -43,6 +45,18 @@ const studentSchema = new Schema({
         type: Number,
         min: 0,
         max: 100
+    },
+    social: {
+        enum:socialgrades,
```
Comment (edenbs) [inline server/api/student/student.model.js]: Does it work with enum AND type number? isn't type number sufficient?

**Review by edenbs — COMMENTED** [review]
Overall looks really good :)
The only big issue is that I'm thinking maybe we should use autocomplete instead of a regular input for the student preferations, it will be easier for teachers to look for students instead of entering their id. Meaning that in the input they will enter the name (the autocomplete will show them all the students filtered by the name and their id in brackets), and when they select the student, it will still show the student name but the ng-model value will be the _id.
Tell me what you think about it

https://material.angularjs.org/latest/demo/autocomplete

File: server/api/student/student.model.js (hunk 4 above)
Comment (noy248) [reply server/api/student/student.model.js]: Right  the type should be enough with the min,max attributes because its numeric.
I deleted the enum reference for the next commit.

File: client/app/students/new-student/new-student.html (hunk 3 above)
Comment (noy248) [reply client/app/students/new-student/new-student.html]: changed at new-student.html and in the edit option at student.html.

File: client/app/students/new-student/new-student.html (hunk 1 above)
Comment (noy248) [reply client/app/students/new-student/new-student.html:8]: It was to separate the form from the save button, without it when adding new values to fill in creating new student the save button will stay in the middle of the form.

File: client/app/students/students.controller.js [hunk 5]
```diff
@@ -132,6 +132,10 @@ angular.module('classify').controller('StudentsController', function($scope, $md
         })
     };
 
+    $scope.isEditor = function () {
+            return auth.hasRole('editor');
```
Comment (edenbs) [inline client/app/students/students.controller.js:144]: why does the indents look weird?

File: client/app/students/students.html [hunk 6]
```diff
@@ -84,7 +84,12 @@
                     <td md-cell ng-if="isEditor()" ng-click="edit($event, item, 'avgGrade')">{{ item.avgGrade }}</td>
                     <td md-cell ng-if="!isEditor()">{{ item.avgGrade }}</td>
                     <td md-cell ng-if="!isEditor()">{{ item.social }}</td>
-                    <td md-cell ng-if="isEditor()" ng-click="edit($event, item, 'social')" >{{ item.social }} </td>
+                    <td md-cell ng-if="isEditor()">
+                        <md-input-container>
```
Comment (edenbs) [inline client/app/students/students.html]: You should use edit inside the cell rather than put a simple input.
See that you can add your constraints like min max in the controller inside the $scope.validators (I see that you did, you should use it with the edit function)

File: client/app/students/students.controller.js [hunk 7]
```diff
@@ -136,7 +146,29 @@ angular.module('classify').controller('StudentsController', function($scope, $md
             });
     };
 
-    $scope.isEditor = function () {
-        return auth.hasRole('editor');
+    $scope.changeSocial = function (student) {
```
Comment (edenbs) [inline client/app/students/students.controller.js]: see comment below -
I think this is unnecessary. edit should be enough :)

File: client/app/students/students.html [hunk 8]
```diff
@@ -80,6 +83,17 @@
                     <td md-cell ng-if="!isEditor()">{{ item.gender }}</td>
                     <td md-cell ng-if="isEditor()" ng-click="edit($event, item, 'avgGrade')">{{ item.avgGrade }}</td>
                     <td md-cell ng-if="!isEditor()">{{ item.avgGrade }}</td>
+                    <td md-cell ng-if="!isEditor()">{{ item.social }}</td>
+                    <td md-cell ng-if="isEditor()">
+                        <md-input-container>
```
Comment (edenbs) [inline client/app/students/students.html]: You should use edit inside the cell rather than put a simple input.
See that you can add your constraints like min max in the controller inside the $scope.validators (I see that you did, you should use it with the edit function)

File: client/index.html [hunk 9]
```diff
@@ -12,11 +12,14 @@
     <link rel="stylesheet" href="bower_components/angular-hamburglar/dist/hamburglar.min.css" />
     <link rel="stylesheet" href="bower_components/mdi/css/materialdesignicons.css" />
     <link rel="stylesheet" href="bower_components/mdi/css/materialdesignicons.min.css" />
+    <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
```
Comment (edenbs) [inline client/index.html]: This shouldn't be inside the bower tag, move it two lines below

File: package.json [hunk 10]
```diff
@@ -56,7 +57,7 @@
     "morgan": "^1.9.0",
     "passport-local-mongoose": "^4.4.0",
     "pify": "^3.0.0",
-    "xlsx": "^0.12.7",
+    "xlsjs": "^1.0.1",
```
Comment (edenbs) [inline package.json:60]: On npm it says the package is deprecated and we should use xlsx, use it if possible
https://www.npmjs.com/package/xlsjs

File: server/api/student/index.js [hunk 11]
```diff
@@ -2,11 +2,13 @@ import {AsyncRouter} from 'express-async-router';
 import * as controller from './student.controller';
 import {hasRole} from '../../auth/auth.service';
 import paginate from 'express-paginate';
-
+import multiparty from 'connect-multiparty';
+var multipartyMiddleware = multiparty();
```
Comment (edenbs) [inline server/api/student/index.js]: Use const for variables that are not gonna change
In ES6 in general we don't use `var` but  `const` and `let`

File: server/api/student/student.model.js [hunk 12]
```diff
@@ -43,6 +43,17 @@ const studentSchema = new Schema({
...
         min: 0,
         max: 100
+    },
+    social: {
+        type: Number,
+        min: 1,
+        max: 4
+    },
+    prefer: {
+        type:{
```
Comment (edenbs) [inline server/api/student/student.model.js:53]: If the label was meant to represent hard coded 'first', 'second'...
I think it's better to actually create the prefer object like this:
```js
prefer: {
  first: {
    id: String
  },
 second: {
  },
 third: {
  }
}
```

File: client/app/students/students.html (hunk 8 above)
Comment (noy248) [reply client/app/students/students.html]: The constraint of the value allow user to get every value on the input, but min/max  
doesn't give the option to other than 1-4.
